File: cpp/src/ch04/auto_move_maze_state.cc
#include <sstream>
#include <random>
#include "auto_move_maze_state.h"

AutoMoveMazeState::AutoMoveMazeState(const int seed) : turn_(0),
                                                       game_score_(0),
                                                       evaluated_score_(0)
{
    auto mt_for_construct = std::mt19937(seed);
    for (int y = 0; y < H; y++)
    {
        for (int x = 0; x < W; x++)
        {
            points_[y][x] = mt_for_construct() % 9 + 1;
        }
    }
}

void AutoMoveMazeState::set_character(const int character_id,
                                      const int y,
                                      const int x)
{
    this->characters_[character_id].y_ = y;
    this->characters_[character_id].x_ = x;
}

bool AutoMoveMazeState::is_done() const
{
    return this->turn_ == END_TURN;
}

std::string AutoMoveMazeState::to_string() const
{
    std::stringstream ss;
    ss << "turn:\t" << this->turn_ << "\n";
    ss << "score:\t" << this->game_score_ << "\n";
    auto board_chars = std::vector<std::vector<char>>(H, std::vector<char>(W, '.'));
    for (int h = 0; h < H; h++)
    {
        for (int w = 0; w < W; w++)
        {
            bool is_written = false; // この座標に書く文字が決定したか

            for (const auto &character : this->characters_)
            {
                if (character.y_ == h && character.x_ == w)
                {
                    ss << "@";
                    is_written = true;
                    break;
                }
                board_chars[character.y_][character.x_] = '@';
            }
            if (!is_written)
            {
                if (this->points_[h][w] > 0)
                {
                    ss << points_[h][w];
                }
                else
                {
                    ss << '.';
                }
            }
        }
        ss << '\n';
    }

    return ss.str();
}

ScoreType AutoMoveMazeState::get_score(bool is_print = false) const
{
    auto tmp_state = *this;
    // remove points on character
    for (auto &character : this->characters_)
    {
        auto &point = tmp_state.points_[character.y_][character.x_];
        point = 0;
    }
    // move till game end
    while (!tmp_state.is_done())
    {
        tmp_state.advance();
        if (is_print)
        {
            std::cout << tmp_state.to_string() << std::endl;
        }
    }
    return tmp_state.game_score_;
}

void AutoMoveMazeState::advance()
{
    for (int character_id = 0; character_id < CHARACTER_N; character_id++)
    {
        move_player(character_id);
    }
    for (auto &character : this->characters_)
    {
        auto &point = this->points_[character.y_][character.x_];
        this->game_score_ += point;
        point = 0; // 被ったら消すため
    }
    this->turn_++;
}

void AutoMoveMazeState::move_player(const int character_id)
{
    Coord &character = this->characters_[character_id];
    int best_point = -INF;
    int best_action_index = 0;

    for (int action = 0; action < 4; action++)
    {
        int ty = character.y_ + dy[action];
        int tx = character.x_ + dx[action];
        if (ty >= 0 && ty < H && tx >= 0 && tx < W)
        {
            auto point = this->points_[ty][tx];
            if (point > best_point)
            {
                best_point = point;
                best_action_index = action;
            }
        }
    }
    character.y_ += dy[best_action_index];
    character.x_ += dx[best_action_index];
}

auto mt_for_action = std::mt19937(0);

void AutoMoveMazeState::init()
{
    for (auto &character : characters_)
    {
        character.y_ = mt_for_action() % H;
        character.x_ = mt_for_action() % W;
    }
}

void AutoMoveMazeState::transition()
{
    auto &character = this->characters_[mt_for_action() % CHARACTER_N];
    character.y_ = mt_for_action() % H;
    character.x_ = mt_for_action() % W;
}

void play_game(const StringAIPair &ai, const int seed)
{
    using std::cout;
    using std::endl;
    auto state = State(seed);
    state = ai.second(state);
    cout << state.to_string() << endl;
    auto score = state.get_score(true);
    cout << "Score of " << ai.first << ": " << score << endl;
}

File: python/src/thunder_book/ch04/auto_move_maze_state.py
from __future__ import annotations

import copy
import enum
import random
from typing import Callable

import numpy as np
from pydantic import BaseModel


class MazeParams(BaseModel):
    width: int
    height: int
    end_turn: int
    num_characters: int


class Coord(BaseModel):
    x: int = 0
    y: int = 0


class D(enum.Enum):
    dx = [1, -1, 0, 0]
    dy = [0, 0, 1, -1]


class MazeState:
    def __init__(self, seed: int, params: MazeParams) -> None:
        self.params = params
        self.points = np.ndarray((params.height, params.width))
        self.turn = 0
        self.characters: list[Coord] = [Coord() for _ in range(params.num_characters)]
        self.game_score = 0
        self.evaluated_score = 0

        self._init_maze(seed)

    def _init_maze(self, seed: int) -> None:
        random.seed(seed)
        for y in range(self.params.height):
            for x in range(self.params.width):
                self.points[y][x] = random.randint(0, 9)

    def init_characters(self) -> None:
        for character in self.characters:
            character.y = random.randint(0, self.params.height - 1)
            character.x = random.randint(0, self.params.width - 1)

    def transition(self) -> None:
        character = random.choice(self.characters)
        character.y = random.randint(0, self.params.height - 1)
        character.x = random.randint(0, self.params.width - 1)

    def set_character(self, character_id: int, y: int, x: int) -> None:
        self.characters[character_id].y = y
        self.characters[character_id].x = x

    def copy(self) -> MazeState:
        return copy.deepcopy(self)

    def get_score(self, should_print: bool = False) -> int:
        tmp_state = self.copy()
        for character in self.characters:
            tmp_state.points[character.y][character.x] = 0
        while not tmp_state.is_done():
            tmp_state.advance()
            if should_print:
                print(tmp_state)
        return tmp_state.game_score

    def __str__(self) -> str:
        map = f"turn: {self.turn}\n"
        map += f"score: {self.game_score}\n"

        for y in range(self.params.height):
            map += "\n"
            for x in range(self.params.width):
                is_written = False
                for character in self.characters:
                    if y == character.y and x == character.x:
                        map += "@"
                        is_written = True
                        break
                if not is_written:
                    if self.points[y][x] > 0:
                        map += f"{int(self.points[y][x])}"
                    else:
                        map += "."

        map += "\n"
        return map

    def is_done(self) -> bool:
        return self.turn >= self.params.end_turn

    def move_player(self, character_id: int) -> None:
        character = self.characters[character_id]
        best_point = -np.inf
        best_action = 0
        # greedy
        for action in range(4):
            ty = character.y + D.dy.value[action]
            tx = character.x + D.dx.value[action]
            if 0 <= ty < self.params.height and 0 <= tx < self.params.width:
                if best_point < self.points[ty][tx]:
                    best_point = self.points[ty][tx]
                    best_action = action
        character.y += D.dy.value[best_action]
        character.x += D.dx.value[best_action]

    def advance(self) -> None:
        for i in range(self.params.num_characters):
            self.move_player(i)

        for character in self.characters:
            point = self.points[character.y][character.x]
            self.game_score += point
            self.points[character.y][character.x] = 0

        self.turn += 1


def play_game(name: str, action_func: Callable[[MazeState], MazeState], seed: int):
    params = MazeParams(width=5, height=5, end_turn=4, num_characters=3)
    state = MazeState(seed, params)
    end_state = action_func(state)
    score = end_state.get_score(should_print=True)
    print(f"Score of {name} is {score}")
